### PR TITLE
Cleanup types [DOC-12]

### PR DESCRIPTION
**Pull request by majakusber**
This is a follow-up PR to https://github.com/diffblue/cbmc/pull/2771 (it is based on top of it so please ignore the `Add documentation` commits until it's merged and rebased).

### CODE REVIEW

**Comment by tautschnig** [comment]
It's a bit of a daring ask, but would you mind tackling #2780, #2808, and possibly even #2810 as well? The last one is likely taking it a bit too far, but the first two might be reasonable? Thank you!!!

File: src/util/type.cpp
Comment (tautschnig) [inline src/util/type.cpp]: I think (and this would make me rather happy) `std_types.h` could go as well? That finally resolves a dependency that should not have existed in the first place...

File: jbmc/src/java_bytecode/ci_lazy_methods_needed.h
Comment (tautschnig) [inline jbmc/src/java_bytecode/ci_lazy_methods_needed.h]: Nit pick: please sort lexicographically.

File: src/util/std_types.cpp
Comment (tautschnig) [inline src/util/std_types.cpp]: As above: please sort lexicographically, likely making a slightly bigger change than necessary (sort all of them).

File: src/util/std_expr.h
Comment (tautschnig) [inline src/util/std_expr.h]: Nit pick: maintain/establish lexicographic sorting

File: src/util/c_types.h
Comment (tautschnig) [inline src/util/c_types.h]: I guess that's really just for `complex_typet`? Maybe keep that in `std_types.h`?

**Review by tautschnig — COMMENTED** [review]
I think this is great cleanup work, though it may be better to do some of this in several small PRs for work not to be held up in case some of it turns out to be controversial.

File: src/util/c_types.h
Comment (majakusber) [reply src/util/c_types.h]: It's a mathematical type so if we're moving some of them into a separate file, it only makes sense if we really move all of them. Is there any particular reason why this stands out for you?

**Comment by majakusber** [comment]
@tautschnig We only had one day booked for doc changes and I'm already out of time :) I'll make sure those issues are put on the next doc day to-do list.

**Comment by majakusber** [comment]
I've split out two of the changes into separate PRs:
- https://github.com/diffblue/cbmc/pull/2820
- https://github.com/diffblue/cbmc/pull/2821

File: src/util/c_types.h
Comment (tautschnig) [reply src/util/c_types.h]: I thought it was the one that's actually used by C/C++ front-ends, but maybe it's not even the only one.

**Comment by tautschnig** [comment]
> We only had one day booked for doc changes and I'm already out of time :)

I know, I'm greedy :-) Thanks a lot for all the work!

**Comment by majakusber** [comment]
Test-gen pointer bump: https://github.com/diffblue/test-gen/pull/2192

**Review by peterschrammel — APPROVED** [review]

**Comment by kroening** [comment]
I find the can_cast_type<symbol_typet>(type1) idiom to be hard to read, and it doesn't add value beyond the one-off. We should not do that.

**Comment by tautschnig** [comment]
In response to @kroening: see ba01589dfd from #2595, which in turn was a side track of #2205.

**Comment by majakusber** [comment]
Yes, the comment in guidelines was exactly my motivation to add missing cast checks.

**Comment by majakusber** [comment]
There is a failing test on the test-gen pointer bump, I will investigate it later. Until then I added `do not merge` label.

**Comment by kroening** [comment]
@tautschnig  Yes; I understand the motivation for can_cast.  However, I'd still like to stick to the if(x.id() ==ID_Z) idiom. In particular, we can then eventually do switch(x.id()).

File: src/util/c_types.h
Comment (majakusber) [reply src/util/c_types.h]: You are right, there is `std::complex` type, moving back to `std_types`.

**Comment by majakusber** [comment]
@kroening Introducing `can_cast`s does not really prevent to access `id` directly. On top of the motivation that was already mentioned, I recently learned that `can_cast` is used by `type/expr_try_dynamic_cast`  which can be quite handy as it can be used in `if` conditions and directly assign the result - thus replacing two lines of can_cast+cast by one (see the third commit in this PR).

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3ced6c7792880b8422c70c1b40e2b2d619c25e94).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84437357
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3ced6c7792880b8422c70c1b40e2b2d619c25e94).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84452470

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 1a28613b3bdaa25060c5f125a55504d6ea670a62).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84984878

**Review by romainbrenguier — APPROVED** [review]

**Pull request merged**
